skyrl-train/skyrl_train/workers/worker.py

```diff
@@ -714,39 +714,41 @@ def optim_step(self) -> float:
 
     def ppo_train(self, train_data: TrainingInputBatch) -> TrainingOutputBatch:
         global_step = train_data.metadata["global_step"]
-        dataloader = BatchIterator(
-            train_data, sample_batch_size=self.cfg.trainer.micro_train_batch_size_per_gpu, drop_last=False
+        minibatch_iterator = BatchIterator(
+            train_data, sample_batch_size=self.policy_mini_batch_size_per_gpu, drop_last=False
         )
 
-        micro_batches_per_mini_batch = (
-            self.policy_mini_batch_size_per_gpu // self.cfg.trainer.micro_train_batch_size_per_gpu
-        )
-        # The number of steps (over micro batches) to accumulate gradients before taking an optimizer step.
-        accumulation_steps = micro_batches_per_mini_batch
-
         status_list = []
         all_metrics = defaultdict(list)
-        policy_update_steps = 0
+        num_minibatches = len(minibatch_iterator)
 
         for epoch in range(self.cfg.trainer.update_epochs_per_batch):
-            pbar = tqdm(
-                dataloader,
+            minibatch_pbar = tqdm(
+                minibatch_iterator,
                 desc=f"Policy Train epoch [{epoch + 1}/{self.cfg.trainer.update_epochs_per_batch}]",
                 disable=not self.strategy.is_rank_0(),
             )
-            for local_step, experience in enumerate(pbar):
-                status = self.forward_backward(experience, accumulation_steps)
+            for local_step, minibatch in enumerate(minibatch_pbar):
+                microbatch_iterator = BatchIterator(
+                    minibatch, sample_batch_size=self.cfg.trainer.micro_train_batch_size_per_gpu, drop_last=False
+                )
+                num_microbatches = len(microbatch_iterator)
+                microbatch_weight = 1.0 / num_microbatches
+
+                for microbatch in microbatch_iterator:
```

Comment:
The next step is basically just to change this microbatch iterator from one that's doing sample-based chunking to one that's token-based.